Breaking change: IsTrimmable for netstandard libs

sbomer opened the pull request:
Adds a breaking change notice about the new behavior of `IsTrimmable` for netstandard libraries.

Fixes https://github.com/dotnet/sdk/issues/30948

Note: I only described the behavior that is shipping in RC1, and listed RC1 as the version that introduced the change. P7 actually introduced the first breaking change, but with slightly different behavior (different code, different message, and it was an error instead of a warning).

<!-- PREVIEW-TABLE-START -->

---

#### Internal previews

| 📄 File | 🔗 Preview link |
|:--|:--|
| [docs/core/compatibility/8.0.md](https://github.com/dotnet/docs/blob/5423d3e2ca24642a82d32575ea1d17d18c3fd04f/docs/core/compatibility/8.0.md) | [Breaking changes in .NET 8](https://review.learn.microsoft.com/en-us/dotnet/core/compatibility/8.0?branch=pr-en-us-36775) |
| [docs/core/compatibility/sdk/8.0/trimming-unsupported-targetframework.md](https://github.com/dotnet/docs/blob/5423d3e2ca24642a82d32575ea1d17d18c3fd04f/docs/core/compatibility/sdk/8.0/trimming-unsupported-targetframework.md) | [docs/core/compatibility/sdk/8.0/trimming-unsupported-targetframework](https://review.learn.microsoft.com/en-us/dotnet/core/compatibility/sdk/8.0/trimming-unsupported-targetframework?branch=pr-en-us-36775) |


<!-- PREVIEW-TABLE-END -->

sbomer committed:
Breaking change: IsTrimmable for netstandard libs

sbomer committed:
Fix lint

gewarren commented on docs/core/compatibility/toc.yml:
```diff
@@ -126,6 +126,8 @@ items:
         href: sdk/8.0/msbuild-language.md
       - name: Runtime-specific apps not self-contained
         href: sdk/8.0/runtimespecific-app-default.md
+      - name: "Warning for IsTrimmable with netstandard libraries"
```
Please also add an entry to line 1469 (one group is by area, one is by version).

gewarren commented on docs/core/compatibility/sdk/8.0/istrimmable-netstandard-libraries.md:
```diff
@@ -0,0 +1,48 @@
+
+## Previous behavior
+
+Previously, `<IsTrimmable>true</IsTrimmable>` would embed an assembly-level attribute `[assembly: AssemblyMetadata("IsTrimmable", "true")]` into the library, opting it into trimming when consumed in a trimmed app (even an app that uses `<TrimMode>partial</TrimMode>`).
```
```suggestion
Previously, `<IsTrimmable>true</IsTrimmable>` would embed an assembly-level attribute `[assembly: AssemblyMetadata("IsTrimmable", "true")]` into the library. That attribute opted the assembly into trimming when consumed in a trimmed app (even an app that uses `<TrimMode>partial</TrimMode>`).
```

gewarren commented on docs/core/compatibility/sdk/8.0/istrimmable-netstandard-libraries.md:
```diff
@@ -0,0 +1,48 @@
+
+## New behavior
+
+In the .NET 8 SDK, building a netstandard library with `<IsTrimmable>` or `<EnableTrimAnalyzer>` produces a warning that `<IsTrimmable>` is unsupported for the target framework, and does not otherwise change the build output.
```
```suggestion
Starting in the .NET 8 SDK, building a netstandard library with `<IsTrimmable>` or `<EnableTrimAnalyzer>` produces a warning that `<IsTrimmable>` is unsupported for the target framework, and does not otherwise change the build output.
```

gewarren commented on docs/core/compatibility/sdk/8.0/istrimmable-netstandard-libraries.md:
```diff
@@ -0,0 +1,48 @@
+
+## Reason for change
+
+Allowing the use of `<IsTrimmable>` or `<EnableTrimAnalyzer>` with netstandard meant that it was easy for library authors to opt into trimming, without being alerted about trim incompatibilities. Because the netstandard reference assemblies were not annotated for trimming, there would be no warnings about uses of framework APIs that are incompatible with trimming.
```
```suggestion
Allowing the use of `<IsTrimmable>` or `<EnableTrimAnalyzer>` with netstandard meant that it was easy for library authors to opt into trimming, without being alerted about trim incompatibilities. Because the netstandard reference assemblies aren't annotated for trimming, there were no warnings about uses of framework APIs that are incompatible with trimming.
```

gewarren commented on docs/core/compatibility/sdk/8.0/istrimmable-netstandard-libraries.md:
```diff
@@ -0,0 +1,48 @@
+
+## Recommended action
+
+Avoid setting `<IsTrimmable>` or `<EnableTrimAnalyzer>` for libraries which target netstandard. Also avoid setting `<IsAotCompatible>`, which implies the former settings. Instead, multi-target the library to include the latest `TargetFramework`, and enable `<IsTrimmable>` only for the supported target frameworks. This will run the latest version of the trim analyzer, using trim compatibility anno ...
```
```suggestion
Avoid setting `<IsTrimmable>` or `<EnableTrimAnalyzer>` for libraries that target netstandard. Also avoid setting `<IsAotCompatible>`, which implies the former settings. Instead, multi-target the library to include the latest `TargetFramework`, and enable `<IsTrimmable>` only for the supported target frameworks. Setting `<IsTrimmable>` will run the latest version of the trim analyzer using trim-compatibility annotations from the latest version of the framework.
```

gewarren reviewed (approved):
I would probably also put all mentions of "netstandard" in backticks.

vitek-karas commented on docs/core/compatibility/sdk/8.0/istrimmable-netstandard-libraries.md:
```diff
@@ -0,0 +1,48 @@
+---
+title: "Breaking change: IsTrimmable warns when used with netstandard libraries"
+description: Learn about a breaking change in the .NE 8 SDK where netstandard libraries that specify IsTrimmable produce a new warning.
```
```suggestion
description: Learn about a breaking change in the .NET 8 SDK where netstandard libraries that specify `IsTrimmable` produce a new warning.
```

vitek-karas commented on docs/core/compatibility/sdk/8.0/istrimmable-netstandard-libraries.md:
```diff
@@ -0,0 +1,48 @@
+---
+# IsTrimmable warns when used with netstandard libraries
+
+Libraries that set `<IsTrimmable>true</IsTrimmable>` with a `TargetFramework` that is any version of netstandard produce a warning that `IsTrimmable` is unsupported.
+Setting `<EnableTrimAnalyzer>true</EnableTrimAnalyzer>` in such a library produces the same warning.
```
I thought the warning is also issued when the library targets NetFX or old .NET Core TFMs, not just netstandard. I think it's OKish to keep the title simple, but the description should be precise - if it's still the case.

sbomer committed:
Apply suggestions from code review

Co-authored-by: Vitek Karas <10670590+vitek-karas@users.noreply.github.com>
Co-authored-by: Genevieve Warren <24882762+gewarren@users.noreply.github.com>

sbomer committed:
Address feedback
- Use .NET Standard instead of netstandard
- Also mention .NET Framework

Also makes the description more thorough by mentioning PublishTrimmed
and PublishAot.

sbomer committed:
Merge remote-tracking branch 'origin/main' into trimmableBreakingChange

sbomer committed:
Fix lint
